### DISCUSSION TITLE
The Zen of Joined Eager Loading and `Result.unique()`

### DISCUSSION THREAD

**Discussion by acarapetis** [sqlalchemy/sqlalchemy#11770] [Usage Questions] [answered]
This is a conceptual question - not really looking for usage help,  just would like to understand a SQLAlchemy2 design choice better.

Here are two quotes from the ORM Querying Guide's Relationship Loading Techniques page.

From *[Joined Eager Loading](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#joined-eager-loading)*:
> When including [joinedload()](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#sqlalchemy.orm.joinedload) in reference to a one-to-many or many-to-many collection, the [Result.unique()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.unique) method must be applied to the returned result, which will uniquify the incoming rows by primary key that otherwise are multiplied out by the join. The ORM will raise an error if this is not present.
> This is not automatic in modern SQLAlchemy, as it changes the behavior of the result set to return fewer ORM objects than the statement would normally return in terms of number of rows. Therefore SQLAlchemy keeps the use of [Result.unique()](https://docs.sqlalchemy.org/en/20/core/connections.html#sqlalchemy.engine.Result.unique) explicit, so there’s no ambiguity that the returned objects are being uniqified on primary key.

And just down the page, from *[The Zen of Joined Eager Loading](https://docs.sqlalchemy.org/en/20/orm/queryguide/relationships.html#the-zen-of-joined-eager-loading)*:
> The strategy can be changed with no other modifications to the query, the results will remain identical, but fewer SQL statements would be emitted. In theory (and pretty much in practice), nothing you can do to the [Select](https://docs.sqlalchemy.org/en/20/core/selectable.html#sqlalchemy.sql.expression.Select) would make it load a different set of primary or related objects based on a change in loader strategy.

To me, it seems these two sections are in conflict: the *Zen* suggests that we should be able to just enable `joinedload` without changing any other code, which was indeed the case in the legacy SQLAlchemy ORM. Now, though, we must call `.unique()` if we are using `joinedload`, but not if we use a different strategy like `lazyload` or `selectinload`.

The main idea that the *Zen* communicates to me is that I *should not* be thinking of a query that contains e.g. `.options(joinedload(User.addresses))` as returning one row per address, but as returning one row per user. So what is the motivation for making the "uniqification" (from row-per-address to row-per-user) be explicit, rather than hiding it as an implementation detail of the `joinedload` strategy?

**Comment by zzzeek** [comment, marked as answer]
> To me, it seems these two sections are in conflict: the Zen suggests that we should be able to just enable joinedload without changing any other code, which was indeed the case in the legacy SQLAlchemy ORM. Now, though, we must call .unique() if we are using joinedload, but not if we use a different strategy like lazyload or selectinload.

you're right, it's a slight conflict. however, you can call .unique() on your statement for those other strategies as well.    that's what Query does, it does a unique() in all cases unconditionally.

> So what is the motivation for making the "uniqification" (from row-per-address to row-per-user) be explicit, rather than hiding it as an implementation detail of the joinedload strategy?

this is an explicit is better than implicit choice, see https://docs.sqlalchemy.org/en/20/changelog/migration_20.html#orm-rows-not-uniquified-by-default  which includes complete discussion

for an example of the classic issue raised by implicit unique see https://github.com/sqlalchemy/sqlalchemy/issues/4395

**Reply by acarapetis** [reply]
thanks for the links :+1:
